Developing-new-features-with-feature-gating.md

```diff
@@ -93,9 +199,37 @@ We use the following principles to determine the value of a feature flags:
+
+### When to use feature flags?
+
+Feature flags should be used when you are working on a feature whose scale makes it hard to implement all at once, or for more experimental features that are likely to cause breakages. Essentially, feature flags are a must for features that are not ready for production/fully-tested yet.
+
+### How to use feature flags?
+
+Say you are working on a large scale user-facing feature that will take multiple PRs to fully implement. In such a case, please use feature flags to gate your feature appropriately by carefully following the steps listed below:
+
+1. The very first PR you make must include changes that add a new feature flag to the codebase, meant to be used with this new feature. The feature flag must be placed in the DEV stage. Every single user-facing aspect of the feature -- whether frontend or backend -- must be gated behind the feature flag (both in the first PR, and all the following ones as well). This is to ensure that the feature is not visible to the user until it is fully implemented and ready for production.
+
+2. The first PR above must be merged before any following PRs are merged. This is to ensure that the feature flag is available in the codebase for it to be used in the following PRs.
+
+3. The very last PR you make (to finish up the feature you are working on) must include changes that move the feature flag to the TEST stage. This is to ensure that the feature is available in the test environment, and we can feature-test it before it is made available to the users in the production environment. **NOTE: Please test all the changes manually to make sure that the feature works fully end-to-end on your local dev server, before flipping the flag to TEST.**
+
+4. If the feature testing reveals that the feature is not ready for production, you must work on fixing the highlighted issues before proceeding further.
 
-For example, if we want to make a feature flag to be enable only in dev environments, we can set the feature flag as the image below.
+5. Once you receive a go-ahead from the feature testers, you must merge another PR -- this PR is meant to do only one thing, i.e. move the feature flag to the PROD stage, allowing it to be enabled/disabled in production (by the admin(s)). **NOTE: When opening this PR, include a link to the testing doc or other proof that the feature has been approved for release.**
 
-![Example feature setting](https://i.imgur.com/GnFQ7El.jpg)
\ No newline at end of file
+6. Once the feature is confirmed to be functioning as intended in production (for at least two weeks), you must merge one last PR to essentially "un-gate" the feature and move the feature flag to the deprecated stage (one of the stages listed in `core/domain/platform_parameter_list.py` for flags that are no longer in use). Then feel free to remove all references of the feature flag from the codebase (for example, in all the `if` blocks you created to gate the feature).
```

Comment:
'Two weeks' feels a bit arbitrary. Shouldn't it be after landing criteria are met and no regressions are detected via stackdriver and other usage metrics? This may vary from feature to feature. @seanlip WDYT?

Reply:
Actually I think two weeks is OK for most things. The idea is to have it running in prod for at least two weeks without issues which I think is reasonable, and I think there should be a rough timeframe here to set expectations.

But to your point, we can include a definition of what "functioning as intended" means. Perhaps something like this:

```
Once the feature is confirmed to be functioning as intended in production (for at least two weeks), please do the following:

- Make sure that the feature is really functioning as intended, and ready to be made permanent. To do this, confirm with the PMs that no users have reported issues with it, and that no regressions have been detected via StackDriver and other usage metrics.
- Once you have confirmation that the feature can be made permanent, merge one last PR to "un-gate" the feature and move the feature flag to the deprecated stage (one of the stages listed in `core/domain/platform_parameter_list.py` for flags that are no longer in use). In that PR, please also remove all remaining references to the feature flag from the codebase (for example, in all the `if` blocks you created to gate the feature).
```

@nithinrdy does this sound good to you?

Reply:
Yep, rewrote instruction 6 with these steps.

Reply:
@seanlip I'm not sure that I understand your reasoning--it actually seems equally arbitrary to the original phrasing. Given that some features might be less discoverable or not on the main happy path, is 2 weeks *actually* enough time to be confident in the feature not causing regressions?

Reply:
To be clear, I'm thinking of 2 weeks as the minimum, not the maximum. This doesn't guarantee that 2 weeks is enough time. It's saying that any feature should be in prod for at least that long. The _actual_ criteria is that "the feature works correctly".

So in the scenario you describe, I imagine that the developer might ask the PM after 2 weeks "are we confident?" and the PM would say "not yet, wait a bit longer". That's WAI.

Do you see an issue with this approach?

Reply:
I think if that's the intent, then it should be codified as such. Based on that, I suggest rewording to:

``... confirmed with the product team to be functioning as intended in production (for no fewer than two weeks)``

Or something similar--WDYT @seanlip @nithinrdy?

Reply:
Hi @BenHenning ... I think that's fine, though I'm not really sure how "at least 2 weeks" is different from "no fewer than 2 weeks". Aren't they logically/mathematically the same? I'm happy with either wording though, if this one is preferred.

Reply:
They're equivalent, but for some reason I read "at least 2 weeks" as "test for 2 weeks" rather than "test for 2 weeks **minimum**". The verbiage change here is meant to make the point a bit stronger, though if this is a me-only interpretation feel free to ignore that suggestion. The "confirmed with the product team" is the more important addition.

Reply:
Sounds good! They mean almost the same thing, although I suppose it's good to be as verbose as possible since the idea is to have the contributors follow this workflow to a tee; the clearer the instructions, the better a contributor can stick to them.